src/components/Firestore/Rules/rules_evaluation_result_model.ts

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** The authentication object as seen by rules */
+export interface RulesAuthObject {
+  uid: string;
+  token: {
+    email?: string;
+    email_verified?: boolean;
+    phone_number?: string;
+    name?: string;
+    sub: string;
+    firebase: {
+      identities: Record<string, string>;
+      sign_in_provider: string;
+    };
+  };
+}
+
+/** The generic rules request */
+export interface RulesRequest {
+  auth?: RulesAuthObject;
+  method: string;
+  path: string;
+  time: number;
+}
+
+/** A Firestore resource that is pulled into the rules context */
+export interface FirestoreResource {
+  __name__: string;
+  id: string;
+  data: Record<string, any>;
+}
+
+/** This corresponds to the `request` variable in rules */
+export interface FirestoreRulesRequest extends RulesRequest {
+  query?: {
+    limit?: number;
+    offset?: number; // Not sure the type for this
+    orderBy?: string;
+  };
+  resource?: FirestoreResource;
+}
+
+/** Everything that's in scope when rules are evaluated */
+export interface FirestoreRulesContext {
+  request: FirestoreRulesRequest;
+  resource?: FirestoreResource;
+}
+
+export type RulesOutcome = 'allow' | 'deny' | 'error';
+
+export interface FirestoreRulesIssue {
+  description: string;
+  severity: string;
+  line: number;
+  col: number;
+}
+export interface FirestoreRulesUpdateData {
+  isCompilationSuccess: boolean;
+  issues: FirestoreRulesIssue[];
+  rules: string;
+}
+
+/** This captures everything about the evaluation, including the outcome */
+export interface FirestoreRulesEvaluation {
+  rulesContext: FirestoreRulesContext;
+  outcome: RulesOutcome;
+  evaluationId: string;
+
+  // TODO: Add a more cleaned-up version of the detailed info
+  granularAllowOutcomes: Array<{
+    line: number;
```

Comment:
I know I wrote this originally, but go ahead and make the outcome type its own interface.
```ts
interface OutcomeInfo { ... }
```

and then use `OutcomeInfo[]` instead of `Array<{...}>`